prov/efa/test/efa_unit_test_ep.c

```diff
@@ -127,3 +127,40 @@ void test_rxr_ep_dc_atomic_error_handling(struct efa_resource **state)
 	/* make sure there is no leaking of tx_entry */
 	assert_true(dlist_empty(&rxr_ep->tx_entry_list));
 }
+
+/**
+ * @brief verify that when shm was used to send small message (<4k), no copy should be perform.
+ *
+ * @param[in]	state		struct efa_resource that is managed by the framework
+ */
+#define BUFF_LEN 8
+void test_rxr_ep_send_with_shm_no_copy(struct efa_resource **state)
+{
+	struct efa_resource *resource = *state;
+	struct efa_ep_addr raw_addr = {0};
+	size_t raw_addr_len = sizeof(struct efa_ep_addr);
+	fi_addr_t peer_addr;
+	int num_addr;
+	char buff[BUFF_LEN] = {0};
+	int err;
+
+	efa_unit_test_resource_construct(resource, FI_EP_RDM);
+
+	/* create a fake peer */
+	err = fi_getname(&resource->ep->fid, &raw_addr, &raw_addr_len);
+	assert_int_equal(err, 0);
+	raw_addr.qpn = 1;
+	raw_addr.qkey = 0x1234;
+	num_addr = fi_av_insert(resource->av, &raw_addr, 1, &peer_addr, 0, NULL);
+	assert_int_equal(num_addr, 1);
+
+	g_ofi_copy_from_hmem_iov_call_counter = 0;
```

Comment:
Move this to mock setup/teardown hook.

Reply:
Also instead of adding a global counter, we can use cmocka mock utilities, i.e. 'will_return_count(efa_mock_ofi_copy_from_hmem_iov_inc_counter, NULL, 0)' - if it was called, cmock will throw an error
https://api.cmocka.org/group__cmocka__mock.html

Reply:
where 
```c
ssize_t efa_mock_ofi_copy_from_hmem_iov_inc_counter(void *dest, size_t size,
						    enum fi_hmem_iface hmem_iface, uint64_t device,
						    const struct iovec *hmem_iov,
						    size_t hmem_iov_count, uint64_t hmem_iov_offset)
{
	mock();
        return 0;
}
```

Reply:
I tried that, but `will_return_count` does not support 0.

Reply:
I see that's a shame.

Then why not even go simpler:

```c
ssize_t efa_mock_ofi_copy_from_hmem_iov_should_not_be_called (void *dest, size_t size,
						    enum fi_hmem_iface hmem_iface, uint64_t device,
						    const struct iovec *hmem_iov,
						    size_t hmem_iov_count, uint64_t hmem_iov_offset)
{
        /* This function should not be called */
        assert(1 == 0);
	return mock();
}
```

Reply:
> Then why not even go simpler:

I tried this too.

This will cause an exception to be throw, while `fi_send` is still holding a lock. 

Then the endpoint cannot be teared down.

Reply:
This is a bit strange... if the test is written correctly this function should not have been invoked at all.

Reply:
> Move this to mock setup/teardown hook.

On this note - it is better than resetting the value manually in the test. The hook is guaranteed to run.

In general, if we use global variables and choose to set/reset its value, we need to make sure the logic always work - if the test fails for some reason and the value is not reset properly, it could cause side effects.

Now you mention that this variable is only used in this 1 test, it is probably ok. But imagine if it is used in multiple tests and it's not reset properly, we could get wrong test results due to the side effect.

Reply:
The test is correct. The libfabric code has a bug which caused the copy function to be called.